nexus/db-model/src/tuf_repo.rs

```diff
@@ -0,0 +1,326 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::str::FromStr;
+
+use crate::{
+    schema::{tuf_artifact, tuf_repo, tuf_repo_artifact},
+    SemverVersion,
+};
+use chrono::{DateTime, Utc};
+use diesel::{deserialize::FromSql, serialize::ToSql, sql_types::Text};
+use omicron_common::{
+    api::external,
+    update::{
+        ArtifactHash as ExternalArtifactHash, ArtifactId as ExternalArtifactId,
+        ArtifactKind,
+    },
+};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use uuid::Uuid;
+
+/// A description of a TUF update: a repo, along with the artifacts it
+/// contains.
+///
+/// This is equivalent to [`external::TufRepoDescription`], but is more
+/// Diesel-friendly.
+#[derive(Debug, Clone)]
+pub struct TufRepoDescription {
+    /// The repository.
+    pub repo: TufRepo,
+
+    /// The artifacts.
+    pub artifacts: Vec<TufArtifact>,
+}
+
+impl TufRepoDescription {
+    /// Creates a new `TufRepoDescription` from an external
+    /// `TufRepoDescription`.
+    ///
+    /// This is not implemented as a `From` impl because we insert new fields
+    /// as part of the process, which `From` doesn't necessarily communicate
+    /// and can be surprising.
+    pub fn from_external(description: external::TufRepoDescription) -> Self {
+        Self {
+            repo: TufRepo::from_external(description.repo),
+            artifacts: description
+                .artifacts
+                .into_iter()
+                .map(TufArtifact::from_external)
+                .collect(),
+        }
+    }
+
+    /// Converts self into [`external::TufRepoDescription`].
+    pub fn into_external(self) -> external::TufRepoDescription {
+        external::TufRepoDescription {
+            repo: self.repo.into_external(),
+            artifacts: self
+                .artifacts
+                .into_iter()
+                .map(TufArtifact::into_external)
+                .collect(),
+        }
+    }
+}
+
+/// A record representing an uploaded TUF repository.
+///
+/// This is equivalent to [`external::TufRepoMeta`], but is more
+/// Diesel-friendly.
```

Comment:
I welcome feedback on this, but the structure of "external" vs "db-model" objects are intentionally distinct. This lets us make database changes without breaking the API.

(They used to all be identical, but this meant that DB refactors -- even column re-ordering -- would result in a breaking API change)

Reply:
Ah yeah that makes sense.